cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/MultiThreadDatasetCache.java

```diff
@@ -0,0 +1,139 @@
+import com.google.common.cache.LoadingCache;
+import com.google.common.cache.RemovalListener;
+import com.google.common.cache.RemovalNotification;
+
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+import javax.annotation.Nullable;
+import javax.annotation.ParametersAreNonnullByDefault;
+
+/**
+ * Implementation of {@link DynamicDatasetCache} that performs all operations on a per-thread basis.
+ * That is, every thread is guaranteed to receive its own distinct copy of every dataset; every thread
+ * has its own transaction context, etc.
+ */
+public class MultiThreadDatasetCache extends DynamicDatasetCache {
+
+  // maintains a single threaded factory for each thread.
+  private final LoadingCache<Long, SingleThreadDatasetCache> perThreadMap;
+
+  /**
+   * See {@link DynamicDatasetCache ).
+   *
+   * @param staticDatasets  if non-null, a map from dataset name to runtime arguments. These datasets will be
+   *                        instantiated immediately, and they will participate in every transaction started
+   *                        through {@link #newTransactionContext()}.
+   */
+  public MultiThreadDatasetCache(final SystemDatasetInstantiator instantiator,
+                                 final TransactionSystemClient txClient,
+                                 final Id.Namespace namespace,
+                                 final Map<String, String> runtimeArguments,
+                                 final MetricsContext metricsContext,
+                                 @Nullable final Map<String, Map<String, String>> staticDatasets) {
+    super(instantiator, txClient, namespace, runtimeArguments, metricsContext);
+    this.perThreadMap = CacheBuilder.newBuilder()
+      .softValues()
+      .removalListener(new RemovalListener<Long, DynamicDatasetCache>() {
+        @Override
+        @ParametersAreNonnullByDefault
+        public void onRemoval(RemovalNotification<Long, DynamicDatasetCache> notification) {
+          DynamicDatasetCache factory = notification.getValue();
+          if (factory != null) {
+            factory.close();
+          }
+        }
+      })
+      .build(
+        new CacheLoader<Long, SingleThreadDatasetCache>() {
+          @Override
+          @ParametersAreNonnullByDefault
+          public SingleThreadDatasetCache load(Long threadId) throws Exception {
+            return new SingleThreadDatasetCache(
+              instantiator, txClient, namespace, runtimeArguments, metricsContext, staticDatasets);
+          }
+        });
+  }
+
+  @Override
+  public void invalidate() {
+    // note that this only invalidates the datasets for the current thread, whereas close() invalidates all
+    entryForCurrentThread().invalidate();
+  }
+
+  @Override
+  public void close() {
+    super.close();
+    perThreadMap.invalidateAll();
+  }
+
+  @Override
+  public synchronized <T extends Dataset> T getDataset(DatasetCacheKey key, boolean bypass)
```

Comment:
no need. I think this is a leftover from the old DynamicDatasetCache (the grandfather of this class) and that had synchronized it.